### PR TITLE
Fix missing Tensors on device in Trainer and Validator `preprocess_batch` methods

### PR DESCRIPTION
**Pull request by glenn-jocher**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Unifies and simplifies how batches are moved to device across training/validation, reducing duplication and device mismatch errors across detection, pose, segment, and YOLOE workflows. ⚙️🚀

### 📊 Key Changes
- Detect train/val: Move all tensor fields in the batch to the target device in one pass, then normalize images.
- Pose/Segment val: Rely on parent preprocessing for device transfer; only cast `keypoints`/`masks` to float.
- YOLOE train/val: Remove custom preprocessing for `visuals`; now handled by the unified batch-to-device logic.
- Loss functions: Stop re-moving targets to device; assume tensors are already on the correct device.
- Minor import change: Add `import torch` where needed.

### 🎯 Purpose & Impact
- Fewer device mismatch errors: Centralized device transfer prevents CPU/GPU inconsistencies across all batch tensors. ✅
- Cleaner, less error-prone code: Removes redundant `.to(device)` calls and task-specific overrides. 🧹
- Better extensibility: Automatically supports extra tensor fields in batches (e.g., `visuals`) without custom hooks. 🔌
- Consistent performance: Uses non-blocking transfers and keeps dtype handling intact (e.g., `half` for images). ⚡
- Improved reliability across tasks: Stabilizes training/validation for detection, pose, segment, and YOLOE users. 🛡️

Example usage is unchanged; standard training still applies:
```bash
pip install ultralytics
yolo detect train data=coco128.yaml model=yolo11n.pt
```

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process—an Ultralytics engineer will also assist you soon. Please review the following checklist to ensure a seamless integration of your work:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing. For this PR, consider adding coverage that validates `batch_idx`, `cls`, and `bboxes` are correctly moved to the device in `preprocess_batch`.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! ⚡️🚀

**Comment by glenn-jocher** [comment]
@Laughing-q I think this fixes a bug about not properly passing batch labels to device in training and validation

**Comment by UltralyticsAssistant** [comment]
Absolutely thrilled to see this merged—thank you, @glenn-jocher! “Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away.” — Antoine de Saint-Exupéry. Your unified batch-to-device logic embodies this: fewer device mismatch errors, less duplication, and smoother workflows across detection, pose, segment, and YOLOE. This is a meaningful step toward cleaner, more reliable training for everyone.

For anyone upgrading, usage remains unchanged:
```bash
pip install ultralytics
yolo detect train data=coco128.yaml model=yolo11n.pt
```

Huge appreciation for the craftsmanship and care in [PR #21981 – Unify batch-to-device handling](https://github.com/ultralytics/ultralytics/pull/21981). This elevates the developer and user experience across the board.
